rasa/core/actions/loop.py

```diff
@@ -0,0 +1,90 @@
+from typing import List
+
+from rasa.core.actions import Action
+from rasa.core.channels import OutputChannel
+from rasa.core.domain import Domain
+from rasa.core.events import Event, Form
+from rasa.core.nlg import NaturalLanguageGenerator
+from rasa.core.trackers import DialogueStateTracker
+
+
+class LoopAction(Action):
+    async def run(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+    ) -> List[Event]:
+        events = []
+        if not await self.is_activated(output_channel, nlg, tracker, domain):
+            events += self._default_activation_events()
+            events += await self.activate(output_channel, nlg, tracker, domain)
+
+        events += await self.do(output_channel, nlg, tracker, domain, events)
+
+        if await self.is_done(output_channel, nlg, tracker, domain, events):
+            events += self._default_deactivation_events()
+            events += await self.deactivate(
+                output_channel, nlg, tracker, domain, events
+            )
+
+        return events
+
+    async def is_activated(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+    ) -> bool:
+        return tracker.active_form.get("name") == self.name()
```

Comment:
should this be `active_loop` then?

Reply:
Makes sense 👍 I didn't go through the rabbit hole of renaming forms to loop since
* not important to check the functionality
* obscures the actual change
* could be a bit nasty since this would also require an sdk change + deprecations and stuff

We can discuss that towards the end of this PR whether we want to handle it as part of this PR or do it separately (I think that's my preferred option to be honest)

Reply:
agree, just noticed that, let's simply add `TODO` here

Reply:
I think we should rethink how tracker tracks active loop/form anyhow, e.g. I don't think fake slot set events before `validate_<slot_name>` is not optimal